src/utils/locks.js

```diff
@@ -0,0 +1,35 @@
+// @flow
+import Package from '../Package';
+import * as npm from './npm';
+
+const LOCK_DIST_TAG = 'pyarn-lock';
+
+export async function lock(packages: Array<Package>) {
+  let locks = [];
+  let promises = [];
+
+  for (let pkg of packages) {
+    let {name, version} = pkg.config;
+    let promise = npm.addTag(name, version, LOCK_DIST_TAG).then(() => {
+      locks.push(pkg);
+    });
+    promises.push(promise);
+  }
+
+  try {
+    await Promise.all(promises);
+  } catch (err) {
+    await unlock(locks);
+    throw new Error('Found a lock tag', err);
```

Comment:
I added the `, err` here because I wanted to show the more detailed error message and wasnt sure whether to move it here, or have it closer to the source of the problem.

Reply:
Oh, silly. I Cant do that.
Do you have an opinion for where this should be handled?
We *need* to catch it higher, but it makes sense to keep the error message closer to where the error occured.
Should this just rethrow the same error after unlocking to you think?